### PR TITLE
ss2: extend for usr context association

### PR DESCRIPTION
**Pull request by cherusk**
Associating pertaining sockets with context holding those via fds.

Signed-off-by: matthias <matthias.tafelmeier@gmx.net>

### CODE REVIEW

**Comment by cherusk** [comment]
Known violations:
cli/ss2:65:9: E731 do not assign a lambda expression, use a def
cli/ss2:68:9: E731 do not assign a lambda expression, use a def

Asking for tolerating those.

File: cli/ss2 [hunk 1]
```diff
@@ -35,12 +38,82 @@ from pyroute2.netlink.diag import (SS_ESTABLISHED,
...
+            self._data[sk_inode][usr][pid]["fds"].append(sk_fd)
+
+        if sk_inode in self._data.keys():
+            if usr in self._data[sk_inode].keys():
+                if pid not in self._data[sk_inode][usr].keys():
+                    _enter_pid(ctxt)
+                _enter_fd(sk_fd)
+            else:
+                self._data[sk_inode][usr] = {}
+                _enter_pid(ctxt)
```
Comment (svinota) [inline cli/ss2]: Nevertheless, you use it here, and it's not only confusing, but also causes unnecessary variable transitions. Simply to ignore the argument in the function and use a variable with the same name from the upper context.

File: cli/ss2 [hunk 2]
```diff
@@ -35,12 +38,82 @@ from pyroute2.netlink.diag import (SS_ESTABLISHED,
...
+                _enter_fd(sk_fd)
+        else:
+            self._data[sk_inode] = {}
+            self._data[sk_inode][usr] = {}
+            _enter_pid(ctxt)
+            _enter_fd(sk_fd)
```
Comment (svinota) [inline cli/ss2]: So if you believe the lambda usage is needed here (I don't), think about that:
1. If you need to call a lambda several times, one of the possible ways is to store lambdas in a dictionary. Technically it's the same as to define a class with methods. From the bytecode perspective it's almost the same.
2. Every variable used within the lambda should be passed as an argument, no implicit upper context lookups — simply 'cause it works a bit weird in some situations.


Here I would define two additional functions in some context, probably in the global context.

**Comment by svinota** [comment]
I would not tolerate. Please, use ordinary function definitions, as here the lambda usage does not simply pays off.

**Comment by svinota** [comment]
The function definitions may be in any scope, I believe, there is no press about speed optimizations. So it's totally up to you to define them within the method or in some upper scope.

**Comment by cherusk** [comment]
In the end, you're right.

As to the performance, actually, even in this ss2 streams, one person once commented actually that aspect as to closures, getting reallocated, yet, holds true for lambdas all the same:

https://github.com/svinota/pyroute2/pull/485

Will unroll it, thanks so far.

**Comment by svinota** [comment]
Thanks!

**Comment by cherusk** [comment]
Resorted to recursion for the entering. Would even describe it as close to aesthetic by now. Advantages should be obvious.

Foreseeing a performance boost only from parallelizing bottle-neck code paths anyways, but that's preliminary.

**Comment by svinota** [comment]
Looks ok. Thanks a lot!

**Comment by svinota** [comment]
please notice that the test is failing: 
```
======================================================================
ERROR: test_cli.TestTools.test_ss2
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/peet/Projects/pyroute2/tests/general/test_cli.py", line 283, in test_ss2
    tcp_flows_hive[what] = future.result()
  File "/usr/lib/python3.6/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/usr/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/peet/Projects/pyroute2/tests/general/test_cli.py", line 260, in do_ss2
    self.ss2.run(args)
  File "./bin/ss2", line 465, in run
    if args.process:
AttributeError: type object 'args' has no attribute 'process'
```

**Comment by svinota** [comment]
nevermid, fixed.

**Pull request merged**